File: rust/src/streams.rs
use ethers::{
    providers::{Provider, Ws},
    types::{Filter, Log, Transaction, U256, U64},
};
use ethers_providers::Middleware;
use std::sync::Arc;
use tokio::sync::broadcast::Sender;
use tokio_stream::StreamExt;

use crate::utils::calculate_next_block_base_fee;

#[derive(Default, Debug, Clone)]
pub struct NewBlock {
    pub block_number: U64,
    pub base_fee: U256,
    pub next_base_fee: U256,
}

#[derive(Debug, Clone)]
pub enum Event {
    Block(NewBlock),
    PendingTx(Transaction),
    Log(Log),
}

pub async fn stream_new_blocks(provider: Arc<Provider<Ws>>, event_sender: Sender<Event>) {
    let stream = provider.subscribe_blocks().await.unwrap();
    let mut stream = stream.filter_map(|block| match block.number {
        Some(number) => Some(NewBlock {
            block_number: number,
            base_fee: block.base_fee_per_gas.unwrap_or_default(),
            next_base_fee: U256::from(calculate_next_block_base_fee(
                block.gas_used,
                block.gas_limit,
                block.base_fee_per_gas.unwrap_or_default(),
            )),
        }),
        None => None,
    });

    while let Some(block) = stream.next().await {
        match event_sender.send(Event::Block(block)) {
            Ok(_) => {}
            Err(_) => {}
        }
    }
}

pub async fn stream_pending_transactions(provider: Arc<Provider<Ws>>, event_sender: Sender<Event>) {
    let stream = provider.subscribe_pending_txs().await.unwrap();
    let mut stream = stream.transactions_unordered(256).fuse();

    while let Some(result) = stream.next().await {
        match result {
            Ok(tx) => match event_sender.send(Event::PendingTx(tx)) {
                Ok(_) => {}
                Err(_) => {}
            },
            Err(_) => {}
        };
    }
}

pub async fn stream_uniswap_v2_events(provider: Arc<Provider<Ws>>, event_sender: Sender<Event>) {
    let filter = Filter::new().event("Sync(uint112,uint112)");
    let mut stream = provider.subscribe_logs(&filter).await.unwrap();

    while let Some(result) = stream.next().await {
        match event_sender.send(Event::Log(result)) {
            Ok(_) => {}
            Err(_) => {}
        };
    }
}

File: javascript/src/streams.js
const { ethers } = require('ethers');

const { calculateNextBlockBaseFee, estimateNextBlockGas } = require('./utils');

function streamNewBlocks(wssUrl, eventEmitter) {
    const wss = new ethers.providers.WebSocketProvider(wssUrl);

    wss.on('block', async (blockNumber) => {
        let block = await wss.getBlock(blockNumber);
        let nextBaseFee = calculateNextBlockBaseFee(block);
        let estimateGas = await estimateNextBlockGas(); 

        eventEmitter.emit('event', {
            type: 'block',
            blockNumber: block.number,
            baseFee: BigInt(block.baseFeePerGas),
            nextBaseFee,
            ...estimateGas,
        });
    });

    return wss;
}

function streamPendingTransactions(wssUrl, eventEmitter) {
    const wss = new ethers.providers.WebSocketProvider(wssUrl);
    
    wss.on('pending', async (txHash) => {
        eventEmitter.emit('event', {
            type: 'pendingTx',
            txHash,
        });
    });

    return wss;
}

function streamUniswapV2Events(wssUrl, eventEmitter) {
    // This stream isn't used in the example DEX arb,
    // but is here to demonstrate how to subscribe to events.
    const wss = new ethers.providers.WebSocketProvider(wssUrl);

    const syncEventSelector = ethers.utils.id('Sync(uint112,uint112)');
    const filter = {topics: [syncEventSelector]};

    wss.on(filter, async (event) => {
        eventEmitter.emit('event', event);
    });

    return wss;
}

module.exports = {
    streamNewBlocks,
    streamPendingTransactions,
    streamUniswapV2Events,
};